### PR TITLE
Use Ruff to lint and format Python files

### PR DESCRIPTION
**Pull request by eliotwrobson**
Remake of #9744. Add Ruff. It's a feature because of how alarmingly fast it is.

Disclaimer: I'm not 1000% sure that the new configs are completely correct. This is the main thing to double check (file changes were from autoformatting and to suppress warnings).

This should be merged after https://github.com/PrairieLearn/PrairieLearn/pull/9363 to avoid an annoying rebase, and that PR is very short.

### CODE REVIEW

File: apps/prairielearn/elements/pl-checkbox/pl-checkbox.py [hunk 1]
```diff
@@ -510,13 +510,13 @@ def parse(element_html, data):
     n_submitted = len(submitted_key)
     if n_submitted > max_options_to_select or n_submitted < min_options_to_select:
         if min_options_to_select != max_options_to_select:
-            data["format_errors"][
-                name
-            ] = f"You must select between <b>{min_options_to_select}</b> and <b>{max_options_to_select}</b> options."
+            data["format_errors"][name] = (
+                f"You must select between <b>{min_options_to_select}</b> and <b>{max_options_to_select}</b> options."
+            )
```
Comment (nwalters512) [inline apps/prairielearn/elements/pl-checkbox/pl-checkbox.py:515]: I kinda like this stylistic choice more 👀

File: graders/c/cgrader/cgrader.py [hunk 2]
```diff
@@ -459,7 +459,9 @@ def compile_re(t):
             comment = (
                 ""
                 if len(exp_output) == 1
-                else " one of" if must_match_all_outputs == "any" else " all of"
+                else " one of"
+                if must_match_all_outputs == "any"
+                else " all of"
```
Comment (nwalters512) [inline graders/c/cgrader/cgrader.py:464]: I actually find this whole thing kind of hard to read; could we split this up into consecutive `if`/`elif`? That said, let's maybe do this in a separate PR first. e.g.:

```py
if len(exp_putput) == 1:
    comment = ("")
elif ...
```

(Assuming I even got the first part of that right, which I'm not confident on, because this is so tricky to read lol)

**Review by nwalters512 — COMMENTED** [review]
The way we've typically done reformatting is to first open a separate PR with _just_ the formatting changes, then immediately follow it with a PR that both updates the formatter config and adds the previous commit hash to `.git-blame-ignore-revs`. The changes here are small enough that this might not be strictly necessary.

File: apps/prairielearn/elements/pl-checkbox/pl-checkbox.py (hunk 1 above)
Comment (eliotwrobson) [reply apps/prairielearn/elements/pl-checkbox/pl-checkbox.py:515]: I felt the same, always thought it was weird black split it up the other way.

File: graders/c/cgrader/cgrader.py (hunk 2 above)
Comment (eliotwrobson) [reply graders/c/cgrader/cgrader.py:464]: It might be worth adding type annotations/checking + doing a comprehensive pass through this file. There's a few places where the code is hard to read.

File: images/plbase/python-requirements.txt [hunk 3]
```diff
@@ -1,13 +1,7 @@
 ansi2html==1.9.1
-black==24.3.0
 chevron==0.14.0
 coloraide==3.1.2
 Faker==24.4.0
-flake8-black==0.3.6
-flake8-isort==6.1.1
-Flake8-pyproject==1.2.3
-flake8==7.0.0
-isort==5.13.2
```
Comment (jonatanschroeder) [inline images/plbase/python-requirements.txt:10]: Should we allow these to remain in case an instructor is relying on them being there in server.py? My instinct is no (these are more likely to be used in an autograder, not in question generation or internal grading), but I just want to make sure this is considered.

Comment (eliotwrobson) [reply images/plbase/python-requirements.txt:10]: I thought about this a little bit too, and I think it's super unlikely. I'm not even sure if flake8 or black are importable from a python script without doing something weird, they're really just command line applications.

Comment (nwalters512) [reply images/plbase/python-requirements.txt:10]: We definitely should not try to account for instructors trying to import these packages. Long-term I'd very much like to separate our dev + deployment images so that instructors can't use our dev packages even if they try.

File: graders/c/cgrader/cgrader.py (hunk 2 above)
Comment (jonatanschroeder) [reply graders/c/cgrader/cgrader.py:464]: This segment is rewritten in #9785 for clarity.

Comment (nwalters512) [reply graders/c/cgrader/cgrader.py:464]: Thanks @jonatanschroeder!

File: Makefile [hunk 4]
```diff
@@ -59,7 +59,10 @@ lint-js:
 	@yarn eslint --ext js --report-unused-disable-directives "**/*.{js,ts}"
 	@yarn prettier --check "**/*.{js,ts,mjs,cjs,mts,cts,md,sql,json,yml,html,css}"
 lint-python:
-	@python3 -m flake8 ./
+# TODO double check these two commands together are exactly what we want
+# https://docs.astral.sh/ruff/linter/
```
Comment (nwalters512) [inline Makefile:63]:
```suggestion
```

File: Makefile [hunk 5]
```diff
@@ -69,9 +72,11 @@ format: format-js format-python
 format-js:
 	@yarn eslint --ext js --fix "**/*.{js,ts}"
 	@yarn prettier --write "**/*.{js,ts,mjs,cjs,mts,cts,md,sql,json,yml,html,css}"
+
+# https://docs.astral.sh/ruff/formatter/#sorting-imports
```
Comment (nwalters512) [inline Makefile:76]:
```suggestion
```

**Review by SethPoulsen — APPROVED** [review]
